modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/ContributorsBuilder.es.js

```diff
@@ -0,0 +1,249 @@
+import React from 'react';
+import {
+	conjunctionShape,
+	contributorShape,
+	operatorShape,
+	propertyGroupShape,
+	propertyTypesShape
+} from '../../utils/types.es';
+import {DragDropContext as dragDropContext} from 'react-dnd';
+import {getPluralMessage} from '../../utils/utils.es';
+
+class ContributorBuilder extends React.Component {
+	static propTypes = {
+		contributors: PropTypes.arrayOf(contributorShape),
+		editing: PropTypes.bool.isRequired,
+		emptyContributors: PropTypes.bool.isRequired,
+		membersCount: PropTypes.number,
+		membersCountLoading: PropTypes.bool,
+		onConjunctionChange: PropTypes.func,
+		onPreviewMembers: PropTypes.func,
+		onQueryChange: PropTypes.func,
+		previewMembersURL: PropTypes.string,
+		propertyGroups: PropTypes.arrayOf(propertyGroupShape),
+		supportedConjunctions: PropTypes.arrayOf(conjunctionShape).isRequired,
+		supportedOperators: PropTypes.arrayOf(operatorShape).isRequired,
+		supportedPropertyTypes: propertyTypesShape.isRequired
+	};
+
+	static defaultProps = {
+		contributors: [],
+		membersCount: 0,
+		onConjunctionChange: () => {},
+		onPreviewMembers: () => {},
+		onQueryChange: () => {},
+		membersCountLoading: false
+	};
+
+	constructor(props) {
+		super(props);
+
+		const {contributors, propertyGroups} = props;
+
+		const firstContributorNotEmpty = contributors.find(
+			contributor => contributor.query !== ''
+		);
+
+		const propertyKey = firstContributorNotEmpty
+			? firstContributorNotEmpty.propertyKey
+			: propertyGroups[0].propertyKey;
+
+		this.state = {
+			editingId: propertyKey
+		};
+	}
+
+	_handleCriteriaChange = (criteriaChange, index) => {
+		const {onQueryChange} = this.props;
+		onQueryChange(criteriaChange, index);
+	};
+
+	_handleCriteriaEdit = (id, editing) => {
+		this.setState({
+			editingId: editing ? undefined : id
+		});
+	};
+
+	render() {
+		const {
+			contributors,
+			editing,
+			emptyContributors,
+			membersCount,
+			membersCountLoading,
+			onConjunctionChange,
+			onPreviewMembers,
+			propertyGroups,
+			supportedConjunctions,
+			supportedOperators,
+			supportedPropertyTypes
+		} = this.props;
+
+		const {editingId} = this.state;
+
+		const rootClasses = getCN('contributor-builder-root', {
+			editing
+		});
+
+		return (
+			<div className={rootClasses}>
+				<div className='criteria-builder-section-sidebar'>
+					<CriteriaSidebar
+						onTitleClicked={this._handleCriteriaEdit}
+						propertyGroups={propertyGroups}
+						propertyKey={editingId}
+					/>
+				</div>
+
+				<div className='criteria-builder-section-main'>
+					<div className='contributor-container'>
+						<div className='container-fluid container-fluid-max-xl'>
+							<div className='content-wrapper'>
+								<div className='sheet'>
+									<div className='d-flex flex-wrap justify-content-between mb-4'>
+										<h2 className='sheet-title mb-2'>
+											{Liferay.Language.get('conditions')}
+										</h2>
+										<div className='criterion-string'>
```

Comment:
^ new line above here. Same rule as above, except for HTML and JSX we add new lines between sibling tags.